### PR TITLE
adaptation to oracle checkpoint

### PR DESCRIPTION
**Pull request by cartersz**
### What problem does this PR solve? 
using oracle db as a checkpoint storage

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `tidb-ansible` repository
 - Need to be included in the release note

### CODE REVIEW

File: drainer/checkpoint/oracle_test.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+package checkpoint
+
+import (
+	"database/sql"
+	"github.com/DATA-DOG/go-sqlmock"
+	. "github.com/pingcap/check"
+	"github.com/pingcap/errors"
+)
+
```
Comment (lichunzhu) [inline drainer/checkpoint/oracle_test.go:22]: Why `TestShouldSaveCheckpoint` and `TestShouldUpdateTsMap` are not added?

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

File: drainer/checkpoint/util.go [hunk 2]
```diff
@@ -83,6 +87,23 @@ func genReplaceSQL(sp *MysqlCheckPoint, str string) string {
 	return fmt.Sprintf("replace into %s.%s values(%d, '%s')", sp.schema, sp.table, sp.clusterID, str)
 }
 
+func genCheckTableIsExist2o(sp *OracleCheckPoint) string {
+	tableName := strings.ToUpper(sp.table)
+	return fmt.Sprintf("select table_name from user_tables where table_name='%s'", tableName)
+}
+
+func genCreateTable2o(sp *OracleCheckPoint) string {
+	return fmt.Sprintf("create table %s.%s (clusterID number primary key, checkPoint varchar2(4000))", sp.schema, sp.table)
```
[the remark this answers is not in the document]
Comment (cartersz) [reply drainer/checkpoint/util.go:96]: for genCheckTableIsExist2o, only the value in the column of table_name in user_tables are uppercase. user_tables is system table. other three oracle method is for checkpoint table by using simple query.

File: drainer/checkpoint/util.go [hunk 3]
```diff
@@ -68,7 +68,11 @@ func setDefaultConfig(cfg *Config) {
 		cfg.Db.User = "root"
 	}
 	if cfg.Schema == "" {
-		cfg.Schema = "tidb_binlog"
+		if cfg.CheckpointType == "oracle" {
+			cfg.Schema = cfg.Db.User
```
Comment (lichunzhu) [inline drainer/checkpoint/util.go:72]: emmmmm. Can you explain why for this?

File: drainer/util.go [hunk 4]
```diff
@@ -118,6 +129,17 @@ func GenCheckPointCfg(cfg *Config, id uint64) (*checkpoint.Config, error) {
...
+		case "oracle":
+			checkpointCfg.CheckpointType = cfg.SyncerCfg.DestDBType
+			checkpointCfg.Db = &checkpoint.DBConfig{
+				Host:                cfg.SyncerCfg.To.Host,
+				User:                cfg.SyncerCfg.To.User,
+				Password:            cfg.SyncerCfg.To.Password,
+				Port:                cfg.SyncerCfg.To.Port,
+				TLS:                 cfg.SyncerCfg.To.TLS,
+				OracleServiceName:   cfg.SyncerCfg.To.OracleServiceName,
+				OracleConnectString: cfg.SyncerCfg.To.OracleConnectString,
```
Comment (lichunzhu) [inline drainer/util.go:141]: Can we use a function to simplify this part?

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM
@glorv PTAL

File: drainer/checkpoint/util.go (hunk 3 above)
Comment (cartersz) [reply drainer/checkpoint/util.go:72]: user in oracle like the schema in TiDB

Comment (glorv) [reply drainer/checkpoint/util.go:72]: If oracle has special requirement, I prefer to force user to set the checkpoint schema

**Review by glorv — COMMENTED** [review]
rest LGTM

File: drainer/util_test.go [hunk 5]
```diff
@@ -204,6 +208,7 @@ func (t *checkPointSuite) TestGenOracleCheckPointCfg(c *C) {
 	c.Assert(cp.Db.User, Equals, cfg.SyncerCfg.To.Checkpoint.User)
 	c.Assert(cp.Db.Password, Equals, cfg.SyncerCfg.To.Checkpoint.Password)
 	c.Assert(cp.Db.TLS, Equals, tls)
+	c.Assert(cp.Table, Equals, "new_checkpoint_table")
```
Comment (lichunzhu) [inline drainer/util_test.go:211]: Add a check for `setDefaultConfig`?

**Review by lichunzhu — COMMENTED** [review]
Report an error when `cfg.User != cfg.Schema` or directly set `cfg.Schema = cfg.User`
Rest LGTM

**Review by lichunzhu — APPROVED** [review]

**Review by glorv — APPROVED** [review]

**Pull request merged**